src/theory/strings/output_channel.cpp

```diff
@@ -0,0 +1,448 @@
+              << "  " << vars[i] << " -> " << subs[i] << std::endl;
+        }
+      }
+      sendLemma(d_true, eqs, c);
+      return;
+    }
+    if (Trace.isOn("strings-lemma-debug"))
+    {
+      for (const Node& u : unproc)
+      {
+        Trace("strings-lemma-debug")
+            << "  non-trivial exp : " << u << std::endl;
+      }
+    }
+  }
+  Trace("strings-lemma") << "Strings::Infer " << eq << " from " << eq_exp
+                         << " by " << c << std::endl;
+  Trace("strings-assert") << "(assert (=> " << eq_exp << " " << eq
+                          << ")) ; infer " << c << std::endl;
+  d_pending.push_back(eq);
+  d_pending_exp[eq] = eq_exp;
+  d_keep.insert(eq);
+  d_keep.insert(eq_exp);
+}
+
+bool OutputChannelStrings::sendSplit(Node a, Node b, const char* c, bool preq)
+{
+  Node eq = a.eqNode(b);
+  eq = Rewriter::rewrite(eq);
+  if (eq.isConst())
+  {
+    return false;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  Node lemma_or = nm->mkNode(OR, eq, nm->mkNode(NOT, eq));
+  Trace("strings-lemma") << "Strings::Lemma " << c << " SPLIT : " << lemma_or
+                         << std::endl;
+  d_pending_lem.push_back(lemma_or);
+  sendPhaseRequirement(eq, preq);
+  return true;
+}
+
+void OutputChannelStrings::sendPhaseRequirement(Node lit, bool pol)
+{
+  d_pending_req_phase[lit] = pol;
+}
+
+Node OutputChannelStrings::getRepresentative(Node t)
+{
+  if (d_ee.hasTerm(t))
+  {
+    return d_ee.getRepresentative(t);
+  }
+  return t;
+}
+
+bool OutputChannelStrings::hasTerm(Node a) { return d_ee.hasTerm(a); }
+
+bool OutputChannelStrings::areEqual(Node a, Node b)
+{
+  if (a == b)
+  {
+    return true;
+  }
+  else if (hasTerm(a) && hasTerm(b))
+  {
+    return d_ee.areEqual(a, b);
+  }
+  return false;
+}
+
+bool OutputChannelStrings::areDisequal(Node a, Node b)
+{
+  if (a == b)
+  {
+    return false;
+  }
+  if (hasTerm(a) && hasTerm(b))
+  {
+    Node ar = d_ee.getRepresentative(a);
+    Node br = d_ee.getRepresentative(b);
+    return (ar != br && ar.isConst() && br.isConst())
+           || d_ee.areDisequal(ar, br, false);
+  }
+  Node ar = getRepresentative(a);
+  Node br = getRepresentative(b);
+  return ar != br && ar.isConst() && br.isConst();
+}
+
+void OutputChannelStrings::doPendingFacts()
+{
+  size_t i = 0;
+  while (!hasConflict() && i < d_pending.size())
+  {
+    Node fact = d_pending[i];
+    Node exp = d_pending_exp[fact];
+    if (fact.getKind() == AND)
+    {
+      for (const Node& lit : fact)
+      {
+        bool polarity = lit.getKind() != NOT;
+        TNode atom = polarity ? lit : lit[0];
+        d_parent.assertPendingFact(atom, polarity, exp);
+      }
+    }
+    else
+    {
+      bool polarity = fact.getKind() != NOT;
+      TNode atom = polarity ? fact : fact[0];
+      d_parent.assertPendingFact(atom, polarity, exp);
+    }
+    i++;
+  }
+  d_pending.clear();
+  d_pending_exp.clear();
+}
+
+void OutputChannelStrings::doPendingLemmas()
+{
+  if (!hasConflict())
```

Comment:
In the original code, we don't do that if we don't have pending lemmas. Is this change intentional? It seems like it could make a difference if we have pending phase requirements but no pending lemmas.

Reply:
Yes, the previous code was a bit nonsensical.  Lemmas and phase requirements essentially should be processed in parallel (phase requirements typically say what polarity to decide for literals that appear in lemmas).  The new code makes it so that we don't do anything if we are in conflict.

Reply:
That's fair, thanks for the explanation.